### PR TITLE
qt6.wrapQtAppsHook: remove dependencies: qtbase qtwayland

### LINKED ISSUE #174828: qt6Packages.callPackage should not add qt6.qtdeclarative to dependencies
low priority

qt6.qtdeclarative is a large package, so it should be excluded for non-GUI qt apps ([example](https://github.com/milahu/qt6-project-template/blob/main/src/main-cli.cc))

currently, when i build a package with `qt6Packages.callPackage` and empty buildInputs, it pulls

* qt6.qtbase
* qt6.qtwayland
  * qt6.qtdeclarative
    * qt6.qtshadertools

the problem is wrapQtAppsHook and its deps

https://github.com/NixOS/nixpkgs/blob/271f15a2384a8744d0b25f3c88484d71b577079b/pkgs/development/libraries/qt-6/default.nix#L86-L89

and qtwayland pulls qtdeclarative

```
$ nix why-depends ./nixpkgs#qt6.qtwayland ./nixpkgs#qt6.qtdeclarative
/nix/store/9hwcqcvv8lb38w57y5pjzqxg65d7d1qs-qtwayland-6.3.0
└───/nix/store/wksjqhyckaji17wmzgx0096hikv5bcvi-qtdeclarative-6.3.0
```

the problem is fixed by removing `self.qtwayland.dev` from deps
wrapQtAppsHook should use only the declared buildInputs

disabling wrapQtAppsHook (dontWrapQtApps = true) is no option, as also CLI qt apps need the qt wrapper

qtwayland is there, so that users dont need to add it to buildInputs
assuming that all qt apps are GUI apps

im in favor of making this consistent with cmake
so that all GUI apps must add qtwayland to buildInputs

CLI qt app:

```cmake
find_package(Qt6 REQUIRED COMPONENTS Core)
target_link_libraries(${target_name} PRIVATE Qt6::Core) # = qtbase
```

GUI qt app:

```cmake
find_package(Qt6 REQUIRED COMPONENTS Core Widgets)
target_link_libraries(${target_name} PRIVATE Qt6::Core Qt6::Widgets) # = qtbase qtwayland
```

since #169296 

cc @NickCao @ttuegel

### PR DESCRIPTION
**Pull request by milahu**
###### Description of changes

fix #174828

make dependencies explicit to reduce closure size for cli apps

qt cli apps: `buildInputs = [ qtbase ];`

qt gui apps: `buildInputs = [ qtbase qtwayland ];`

qt wrappers like pyqt or pyside: ?

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
A second thought on this lead me to believe that qtbase thould be removed from deps too. The hook itself does not require qtbase as a dep.

**Comment by NickCao** [comment]
A clean separation between the hook which runs on the buildPlatform, and qtbase on the hostPlatform would also make cross compilations eaiser, without having to deal with splicing and friends.

File: pkgs/development/libraries/qt-6/default.nix [hunk 1]
```diff
@@ -85,7 +85,7 @@ let
 
       wrapQtAppsHook = makeSetupHook
         {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-6/default.nix:87]:
```suggestion
      wrapQtAppsHook = makeSetupHook {
```

**Comment by nrdxp** [comment]
So this doesn't have to advertised affect, and so perhaps this should be reverted, unless/until we find the problem.

**Comment by SuperSandro2000** [comment]
> So this doesn't have the advertised affect, and so perhaps this should be reverted, unless/until we find the problem.

It has at least one effect: It reduces closure size and I would like to keep that.

If qtwayland needs to be added back to the hook, we should think about an extra hook or make qtwayland optional.

**Comment by FlorianFranzen** [comment]
Can we please remove the 'xcb' backends as well, as it bloats all qt derivations for me on wayland. 

As a matter of fact why do we not remove all platform plugin. Per default that seems to be eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc and xcb, so there is still a lot of bloat to remove.

It might force users to rebuild all qt applications if they want to use qt at all, but it makes the derivation very small, which seems the main goal here. It is their fault to choose to run a bloaty graphical user interface to begin with.

**Pull request merged**